torch/csrc/jit/script/object.h

```diff
@@ -28,34 +28,60 @@ struct TORCH_API Object {
   }
 
   void setattr(const std::string& name, c10::IValue v) {
-    size_t slot = _ivalue()->type()->getAttributeSlot(name);
-    const c10::TypePtr& expected = _ivalue()->type()->getAttribute(slot);
-    TORCH_CHECK(expected, "Module has no attribute '", name, "'");
-    TORCH_CHECK(
-        v.type()->isSubtypeOf(expected),
-        "Expected a value of type '",
-        expected->python_str(),
-        "' for field '",
-        name,
-        "', but found '",
-        v.type()->python_str(),
-        "'");
-    _ivalue()->setSlot(slot, std::move(v));
+    if (auto slot = _ivalue()->type()->findAttributeSlot(name)) {
+      const c10::TypePtr& expected = _ivalue()->type()->getAttribute(*slot);
+      TORCH_CHECK(
+          v.type()->isSubtypeOf(expected),
+          "Expected a value of type '",
+          expected->python_str(),
+          "' for field '",
+          name,
+          "', but found '",
+          v.type()->python_str(),
+          "'");
+      _ivalue()->setSlot(*slot, std::move(v));
+    } else if (auto v = _ivalue()->type()->getConstant(name)) {
+      TORCH_CHECK(
+          false,
+          "Can't set constant '",
+          name,
+          "' which has value:",
+          v);
+    } else {
+      TORCH_CHECK(
+          false,
+          "Module has no attribute '", name, "'");
+    }
   }
 
   c10::IValue attr(const std::string& name) const {
-    return _ivalue()->getAttr(name);
+    if (auto r = _ivalue()->type()->findAttributeSlot(name)) {
+      return _ivalue()->getSlot(*r);
+    }
+    if (_ivalue()->type()->hasConstant(name)) {
+      return *_ivalue()->type()->getConstant(name);
```

Comment:
I agree with the naming consistency comment. Our general convention for retrieving `Foo` is to define three functions: 
- `findFoo()` returns a nullable object (like pointer or `optional`) and the user is responsible for checking the null state
- `getFoo()` is a checked retrieval, which throws an exception if the name is not present.
- `hasFoo()` which is just an existence check.

Can we make `ClassType` conform to this convention? Probably in a separate PR stacked under this one.

Reply:
I think we are conforming to this convention already. I'm following attribute for constant APIs
One problem is getAttribute returns a TypePtr which is nullable, and it returns `nullptr` when it fails: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/jit_type.h#L1445, but for constant we don't have `nullptr` that's why I'm using `c10::optional<IValue>` here, or should I change it to return a None IValue?

Reply:
Can't getConstant just return a non-nullable IValue or throw an exception? Don't worry about conforming to `getAttribute` it is wrong as well.

Reply:
yeah I'm going to throw exception in both `getConstant` and `getAttribute` when the field is not found, added two PRs in the stack.